21.1.6+ (#3)
DevExpress-Examples/web-forms-dashboard-extract-data-source: Readme.md

[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T506198)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->

# Dashboard for Web Forms - How to Use DashboardExtractDataSource to provide data

This example demonstrates basic approaches and code snippets that can be used to bind the [ASPxDashboard](https://docs.devexpress.com/Dashboard/DevExpress.DashboardWeb.ASPxDashboard) control to the [DashboardExtractDataSource](https://docs.devexpress.com/Dashboard/DevExpress.DashboardCommon.DashboardExtractDataSource). 

> The [Extract Data Source](https://docs.devexpress.com/Dashboard/115900) improves performance when a complex query or a stored procedure takes a significant time to get data from a database. 

Reviewer: ```suggestion
The [Extract Data Source](https://docs.devexpress.com/Dashboard/115900) improves performance when a complex query or a stored procedure takes a significant amount of time to get data from a database. 
```

> The [DashboardExtractDataSource](https://docs.devexpress.com/Dashboard/DevExpress.DashboardCommon.DashboardExtractDataSource) class implements the Extract Data Source concept and allows you to request the data once and save it in the compressed and optimized form to a file. Subsequently an application can retrieve data from that file or create a new file when data is updated.

Reviewer: ```suggestion
The [DashboardExtractDataSource](https://docs.devexpress.com/Dashboard/DevExpress.DashboardCommon.DashboardExtractDataSource) class implements the Extract Data Source concept, and allows you to request data once and save it in the compressed and optimized form to a file. Subsequently, an application can retrieve data from that file or create a new file when data is updated.
```

> The [DashboardExtractDataSource](https://docs.devexpress.com/Dashboard/DevExpress.DashboardCommon.DashboardExtractDataSource) class implements the Extract Data Source concept and allows you to request the data once and save it in the compressed and optimized form to a file. Subsequently tan application can retrieve data from that file or create a new file when data is updated.

Reviewer 2: ```suggestion
The [DashboardExtractDataSource](https://docs.devexpress.com/Dashboard/DevExpress.DashboardCommon.DashboardExtractDataSource) class implements the Extract Data Source concept and allows you to request the data once and save it in the compressed and optimized form to a file. Subsequently an application can retrieve data from that file or create a new file when data is updated.
```

}
```

> To update the data extract file and load the updated data in ASPxDashboard, send an AJAX request to the server and call the [DashboardExtractDataSource.UpdateFile](https://docs.devexpress.com/Dashboard/DevExpress.DashboardCommon.DashboardExtractDataSource.UpdateFile(DashboardExtractDataSource--Action-String--ExtractUpdateResult---Action-String--ExtractUpdateResult-)) method there. Once a new file is created on the server, you can return the callback back to the client and call the [ASPxClientDashboard.ReloadData](http://docs.devexpress.devx/Dashboard/js-ASPxClientDashboard?p=netframework#js_aspxclientdashboard_reloaddata) method to reload the control with new data:

Reviewer 2: ```suggestion
To update the data extract file and load the updated data in ASPxDashboard, send an AJAX request to the server and call the [DashboardExtractDataSource.UpdateFile](https://docs.devexpress.com/Dashboard/DevExpress.DashboardCommon.DashboardExtractDataSource.UpdateFile(DashboardExtractDataSource--Action-String--ExtractUpdateResult---Action-String--ExtractUpdateResult-)) method there. Once a new file is created on the server, you can return the callback back to the client and call the [ASPxClientDashboard.ReloadData](https://docs.devexpress.com/Dashboard/js-ASPxClientDashboard?p=netframework#js_aspxclientdashboard_reloaddata) method to reload the control with new data:
```

```js
function UpdateExtractDataSource() {
    $.ajax({
	url: "Default.aspx/UpdateExtractDataSource",
	type: "POST",
	data: {},
	contentType: "application/json; charset=utf-8"
    }).then(function (result) {
	dashboard.ReloadData();
	DevExpress.ui.notify(result.d, "success", 5000);
    }, function () {
	DevExpress.ui.notify("We could not update extract data source.", "error", 2000)
    });
}
```

```cs
[WebMethod]
public static string UpdateExtractDataSource() {
    DashboardExtractDataSource ds = CreateExtractDataSource();
    StringBuilder sb = new StringBuilder("We updated your extract data source. ");
    var task = DashboardExtractDataSource.UpdateFile(ds,
	(fileName, result) => {
	    sb.AppendLine($"{DateTime.Now.ToString("T")} - Data Updated - {result} - {Path.GetFileName(fileName)}. ");
	},
	(fileName, result) => {
	    sb.AppendLine($"{DateTime.Now.ToString("T")} - File Updated - {result} - {Path.GetFileName(fileName)}. ");
	});
    // Wait until the data is refreshed in the Extract Data Source.
    task.Wait();
    return sb.ToString();
}
```

> In this example, click a button to extract data. However, in a real-life application, this solution can be insufficient (for example, the site may be deployed to the web farm server). We recommend that you create a separate windows service that should update data automatically every hour or every day.

Reviewer: ```suggestion
In this example, click a button to extract data. However, in a real-life application, this solution can be insufficient (for example, the site may be deployed to the web farm server). We recommend that you create a separate windows service that updates data automatically every hour or every day.
```
